File: backend/internal/payload/tensorboard.go
package payload

// TensorboardSourceJobReq identifies one source job and optionally overrides
// the TensorBoard event directory declared by that job.
type TensorboardSourceJobReq struct {
	JobName string `json:"jobName"`
	LogDir  string `json:"logDir,omitempty"`
}

// CreateTensorboardReq describes the optional source jobs, log directory, and panel lifetime.
// Without a source job, LogDir must point into the current user's personal workspace.
// With source jobs, the backend resolves trusted data mounts from jobs owned by the current user.
// SourceJobName, SourceJobNames, and LogDir remain available for legacy API clients.
type CreateTensorboardReq struct {
	SourceJobName  string                    `json:"sourceJobName,omitempty" example:"job-old-xxxx"`
	SourceJobNames []string                  `json:"sourceJobNames,omitempty"`
	SourceJobs     []TensorboardSourceJobReq `json:"sourceJobs,omitempty"`
	LogDir         string                    `json:"logDir,omitempty" example:"/mnt/vol0/logs"`
	TTLHours       int32                     `json:"ttlHours" binding:"required,min=1,max=168" example:"24"`
}

// TensorboardSourceConfigResp contains TensorBoard settings declared by a source job.
type TensorboardSourceConfigResp struct {
	LogDir string `json:"logDir"`
}

// CreateTensorboardResp describes a newly created TensorBoard panel.
type CreateTensorboardResp struct {
	TensorboardID string `json:"tensorboardId"` // Unique panel identifier.
	AccessPath    string `json:"accessPath"`    // Relative route prefix used to access the panel.
}

type TensorboardStatus string

type TensorboardStatusReason string

const (
	TensorboardStatusStarting TensorboardStatus = "starting"
	TensorboardStatusReady    TensorboardStatus = "ready"
	TensorboardStatusFailed   TensorboardStatus = "failed"

	TensorboardStatusReasonDeploymentFailed        TensorboardStatusReason = "deployment_failed"
	TensorboardStatusReasonStatusCheckPending      TensorboardStatusReason = "status_check_pending"
	TensorboardStatusReasonPodListPending          TensorboardStatusReason = "pod_list_pending"
	TensorboardStatusReasonImagePullFailed         TensorboardStatusReason = "image_pull_failed"
	TensorboardStatusReasonContainerStartFailed    TensorboardStatusReason = "container_start_failed"
	TensorboardStatusReasonContainerExited         TensorboardStatusReason = "container_exited"
	TensorboardStatusReasonDeploymentStarting      TensorboardStatusReason = "deployment_starting"
	TensorboardStatusReasonNetworkConfigIncomplete TensorboardStatusReason = "network_config_incomplete"
	TensorboardStatusReasonServiceMissing          TensorboardStatusReason = "service_missing"
	TensorboardStatusReasonNetworkCheckPending     TensorboardStatusReason = "network_check_pending"
	TensorboardStatusReasonServiceMisconfigured    TensorboardStatusReason = "service_misconfigured"
	TensorboardStatusReasonIngressMissing          TensorboardStatusReason = "ingress_missing"
	TensorboardStatusReasonIngressMisconfigured    TensorboardStatusReason = "ingress_misconfigured"
	TensorboardStatusReasonEndpointPending         TensorboardStatusReason = "endpoint_pending"
	TensorboardStatusReasonReady                   TensorboardStatusReason = "ready"
)

// TensorboardInfo describes a panel and its current availability.
type TensorboardInfo struct {
	ID            string                  `json:"id"`
	Expiration    string                  `json:"expiration"`
	CreatedAt     string                  `json:"createdAt"`
	AccessPath    string                  `json:"accessPath"`
	Status        TensorboardStatus       `json:"status"`
	StatusReason  TensorboardStatusReason `json:"statusReason"`
	StatusMessage string                  `json:"statusMessage"`
}

// ExtendTTLReq resets a TensorBoard panel lifetime from the current time.
type ExtendTTLReq struct {
	TTLHours int32 `json:"ttlHours" binding:"required,min=1,max=168"`
}

File: frontend/src/services/api/tensorboard.ts
import { apiV1Delete, apiV1Get, apiV1Post } from '@/services/client'

export const MAX_TENSORBOARD_SOURCE_JOBS = 10
export const MAX_ACTIVE_TENSORBOARDS = 10

export interface TensorboardSourceJobReq {
  jobName: string
  logDir?: string
}

export interface CreateTensorboardReq {
  sourceJobName?: string
  sourceJobNames?: string[]
  sourceJobs?: TensorboardSourceJobReq[]
  logDir: string
  ttlHours: number
}

export interface CreateTensorboardResp {
  tensorboardId: string
  accessPath: string
}

export type TensorboardStatus = 'starting' | 'ready' | 'failed'

export type TensorboardStatusReason =
  | 'deployment_failed'
  | 'status_check_pending'
  | 'pod_list_pending'
  | 'image_pull_failed'
  | 'container_start_failed'
  | 'container_exited'
  | 'deployment_starting'
  | 'network_config_incomplete'
  | 'service_missing'
  | 'network_check_pending'
  | 'service_misconfigured'
  | 'ingress_missing'
  | 'ingress_misconfigured'
  | 'endpoint_pending'
  | 'ready'

export interface TensorboardInfo {
  id: string
  expiration: string
  createdAt: string
  accessPath: string
  status: TensorboardStatus
  statusReason?: TensorboardStatusReason
  statusMessage: string
}

export interface TensorboardSourceConfig {
  logDir: string
}

export function apiTensorboardList() {
  return apiV1Get<TensorboardInfo[]>('tensorboard')
}

export function apiTensorboardCreate(data: CreateTensorboardReq) {
  return apiV1Post<CreateTensorboardResp>('tensorboard', data)
}

export function apiTensorboardSourceConfig(jobName: string) {
  return apiV1Get<TensorboardSourceConfig>(`tensorboard/source/${encodeURIComponent(jobName)}`)
}

export function apiTensorboardExtendTTL(id: string, ttlHours: number) {
  return apiV1Post<string>(`tensorboard/${id}/extend`, { ttlHours })
}

export function apiTensorboardCreateAccessSession(id: string) {
  return apiV1Post<string>(`tensorboard/${id}/access`)
}

export function apiTensorboardDelete(id: string) {
  return apiV1Delete<string>(`tensorboard/${id}`)
}
